File: ex21/alignImages.py
import argparse
import cv2
import numpy as np
from skimage import io, transform
from skimage.registration import phase_cross_correlation
from skimage.transform import warp, AffineTransform

def detect_and_match_keypoints(reference_image, target_image):
    # Initialize SIFT detector
    sift = cv2.SIFT_create()

    # Detect keypoints and descriptors in both images
    keypoints_ref, descriptors_ref = sift.detectAndCompute(reference_image, None)
    keypoints_target, descriptors_target = sift.detectAndCompute(target_image, None)

    # FLANN-based matcher
    FLANN_INDEX_KDTREE = 1
    index_params = dict(algorithm=FLANN_INDEX_KDTREE, trees=5)
    search_params = dict(checks=50)
    flann = cv2.FlannBasedMatcher(index_params, search_params)

    # Match descriptors
    matches = flann.knnMatch(descriptors_ref, descriptors_target, k=2)

    # Filter matches using the Lowe's ratio test
    good_matches = []
    for m, n in matches:
        if m.distance < 0.7 * n.distance:
            good_matches.append(m)

    # Extract location of good matches
    points_ref = np.float32([keypoints_ref[m.queryIdx].pt for m in good_matches])
    points_target = np.float32([keypoints_target[m.trainIdx].pt for m in good_matches])

    return points_ref, points_target, good_matches

def align_images_sift(reference_image, target_image):
    # Detect keypoints and match them
    points_ref, points_target, good_matches = detect_and_match_keypoints(reference_image, target_image)

    # Compute homography using RANSAC
    homography, mask = cv2.findHomography(points_target, points_ref, cv2.RANSAC, 5.0)

    # Warp the target image using the homography
    height, width = reference_image.shape
    aligned_image = cv2.warpPerspective(target_image, homography, (width, height))

    # Extract translation parameters from homography
    tx, ty = extract_translation(homography)

    return aligned_image, tx, ty

def align_images_phase_correlation(reference_image, target_image):
    # Compute the translation parameters using phase correlation
    shift, error, diffphase = phase_cross_correlation(reference_image, target_image)

    # Apply the translation to the target image
    transform = AffineTransform(translation=(-shift[1], -shift[0]))
    aligned_image = warp(target_image, transform, mode='constant')

    return aligned_image, shift[1], shift[0]

def extract_translation(homography):
    # Normalize the homography matrix
    homography = homography / homography[2, 2]

    # Extract translation parameters
    tx = homography[0, 2]
    ty = homography[1, 2]

    return tx, ty

def main():
    # Parse command line arguments
    parser = argparse.ArgumentParser(description="Align a target image to a reference image.")
    parser.add_argument("-r", "--reference", required=True, help="Path to the reference image.")
    parser.add_argument("-t", "--target", required=True, help="Path to the target image.")
    parser.add_argument("-o", "--output", required=True, help="Path to save the aligned image.")
    parser.add_argument("-a", "--algorithm", type=int, required=True, choices=[0, 1],
                        help="Algorithm to use: 0 for SIFT, 1 for phase correlation.")
    args = parser.parse_args()

    # Load the reference and target images
    reference_image = cv2.imread(args.reference, cv2.IMREAD_GRAYSCALE)
    target_image = cv2.imread(args.target, cv2.IMREAD_GRAYSCALE)

    # Check if images were loaded successfully
    if reference_image is None or target_image is None:
        print("Error: Could not load images.")
        return

    # Align the target image to the reference image using the selected algorithm
    if args.algorithm == 0:
        print("Using SIFT-based homography for alignment.")
        aligned_image, tx, ty = align_images_sift(reference_image, target_image)
    elif args.algorithm == 1:
        print("Using phase correlation for alignment.")
        aligned_image, tx, ty = align_images_phase_correlation(reference_image, target_image)
    else:
        print("Invalid algorithm selection.")
        return

    # Save the aligned image
    cv2.imwrite(args.output, aligned_image)

    # Print the translation parameters
    print(f"Translation parameters (in pixels): (tx, ty) = ({tx:.2f}, {ty:.2f})")

if __name__ == "__main__":
    main()

File: ex21/alignImages.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <string>

using namespace cv;
using namespace std;

// Function to align images using SIFT-based homography
void alignImagesSIFT(const Mat& reference, const Mat& target, Mat& alignedImage, double& tx, double& ty) {
    // Detect keypoints and compute descriptors using SIFT
    Ptr<SIFT> sift = SIFT::create();
    vector<KeyPoint> keypointsRef, keypointsTarget;
    Mat descriptorsRef, descriptorsTarget;
    sift->detectAndCompute(reference, noArray(), keypointsRef, descriptorsRef);
    sift->detectAndCompute(target, noArray(), keypointsTarget, descriptorsTarget);

    // Match descriptors using FLANN
    Ptr<DescriptorMatcher> matcher = DescriptorMatcher::create(DescriptorMatcher::FLANNBASED);
    vector<vector<DMatch>> knnMatches;
    matcher->knnMatch(descriptorsRef, descriptorsTarget, knnMatches, 2);

    // Filter matches using Lowe's ratio test
    vector<DMatch> goodMatches;
    for (size_t i = 0; i < knnMatches.size(); i++) {
        if (knnMatches[i][0].distance < 0.7 * knnMatches[i][1].distance) {
            goodMatches.push_back(knnMatches[i][0]);
        }
    }

    // Extract location of good matches
    vector<Point2f> pointsRef, pointsTarget;
    for (const auto& match : goodMatches) {
        pointsRef.push_back(keypointsRef[match.queryIdx].pt);
        pointsTarget.push_back(keypointsTarget[match.trainIdx].pt);
    }

    // Compute homography using RANSAC
    Mat homography = findHomography(pointsTarget, pointsRef, RANSAC);

    // Warp the target image using the homography
    warpPerspective(target, alignedImage, homography, reference.size());

    // Extract translation parameters from homography
    tx = homography.at<double>(0, 2);
    ty = homography.at<double>(1, 2);
}

// Function to align images using phase correlation
void alignImagesPhaseCorrelation(const Mat& reference, const Mat& target, Mat& alignedImage, double& tx, double& ty) {
    // Compute phase correlation
    Mat hann;
    createHanningWindow(hann, reference.size(), CV_32F);
    Mat referenceFloat, targetFloat;
    reference.convertTo(referenceFloat, CV_32F);
    target.convertTo(targetFloat, CV_32F);
    multiply(referenceFloat, hann, referenceFloat);
    multiply(targetFloat, hann, targetFloat);

    Mat response;
    phaseCorrelate(referenceFloat, targetFloat, response);

    // Extract translation parameters
    tx = response.x;
    ty = response.y;

    // Apply translation to the target image
    Mat translationMatrix = (Mat_<double>(2, 3) << 1, 0, -tx, 0, 1, -ty;
    warpAffine(target, alignedImage, translationMatrix, reference.size());
}

int main(int argc, char** argv) {
    // Parse command-line arguments
    if (argc != 5) {
        cerr << "Usage: " << argv[0] << " -r <reference_image> -t <target_image> -o <output_image> -a <algorithm>" << endl;
        cerr << "Algorithm: 0 for SIFT, 1 for phase correlation" << endl;
        return -1;
    }

    string referencePath = argv[2];
    string targetPath = argv[3];
    string outputPath = argv[4];
    int algorithm = stoi(argv[6]);

    // Load images
    Mat reference = imread(referencePath, IMREAD_GRAYSCALE);
    Mat target = imread(targetPath, IMREAD_GRAYSCALE);

    if (reference.empty() || target.empty()) {
        cerr << "Error: Could not load images." << endl;
        return -1;
    }

    // Align images based on the selected algorithm
    Mat alignedImage;
    double tx = 0, ty = 0;

    if (algorithm == 0) {
        cout << "Using SIFT-based homography for alignment." << endl;
        alignImagesSIFT(reference, target, alignedImage, tx, ty);
    } else if (algorithm == 1) {
        cout << "Using phase correlation for alignment." << endl;
        alignImagesPhaseCorrelation(reference, target, alignedImage, tx, ty);
    } else {
        cerr << "Invalid algorithm selection." << endl;
        return -1;
    }

    // Save the aligned image
    imwrite(outputPath, alignedImage);

    // Print translation parameters
    cout << "Translation parameters (in pixels): (tx, ty) = (" << tx << ", " << ty << ")" << endl;

    return 0;
}